docs/examples/README.md

```diff
@@ -22,7 +22,9 @@ As you can see in `training.ts` a client is represented by a `Disco` object:
 
 ```js
 const disco = new Disco(task, { url, scheme: "federated" });
-await disco.fit(dataset); // Start training on the dataset
+for await (const round of disco.fit(dataset))
+  for await (const epoch of round)
+    for await (const batch of epoch);
```

Comment:
quick question here, why do we have epochs actually? as rounds (of aggr) can often also be much smaller than one epoch. wouldn't it be easiest to only have rounds and batches(=steps)?

slightly related to the LLM also, where it's also fine to train for less than one epoch if time is of concern?

Reply:
> quick question here, why do we have epochs actually? as rounds (of aggr) can often also be much smaller than one epoch. wouldn't it be easiest to only have rounds and batches(=steps)?

historically, it was taken from tfjs as it is working by epochs, so the same relation of "round -> epoch -> batch" was kept.
we can flatten it to have "round -> batch", that does seems easier. won't there be an issue if the dataset is not fully used? we can have batches randomly sampled from the dataset to avoid that, which would make datasets random access rather than linear, require much rework of the loaders. but that's doable.

> slightly related to the LLM also, where it's also fine to train for less than one epoch if time is of concern?

for now, it's hardcoded to only train 5 batches per epoch with gpt-tfjs, as to have a faster result (good for demo).

Reply:
Fyi in tfjs every dataset (image text tabular) is currently already accessed in random batch order rather than linearly by default

Reply:
epoch has a clear meaning, that every element of the dataset is used exactly once.  so better not redefine it to be 5 batches. much easier to make your round 5 batches (or whatever value it already had now, per round)

either remove epoch entirely, or make clear it still has a precise meaning (that a round is 0.2 epochs for example, but don't say an epoch is 5 batches if that's not respecting the meaning). i'm pretty sure at some point we did allow fractional values for epoch so we could keep doing that also. as you prefer

if a training only accesses a small fraction of the data overall, that's fine too. 

if it's random or not is another story, wouldn't matter here either way

Reply:
> epoch has a clear meaning, that every element of the dataset is used exactly once. so better not redefine it to be 5 batches. much easier to make your round 5 batches (or whatever value it already had now, per round)

this redefining to at most 5 batches is only done in gpt-tfjs, to avoid taking too much time training it, the classic tfjs models are using the full dataset. it was done to avoid training for too long.

> either remove epoch entirely, or make clear it still has a precise meaning (that a round is 0.2 epochs for example, but don't say an epoch is 5 batches if that's not respecting the meaning).

I do agree that we should respect the meaning of "epoch". it's quite weird to arbitrary stop. we can remove this limitation of gpt-tfjs. showing the increasing batch count does help user with choosing to reduce the size of the training dataset (wikitext is quite large). it's reduce a bit the usability of the demo tough.

> i'm pretty sure at some point we did allow fractional values for epoch so we could keep doing that also. as you prefer

I was unable to find such fractional epoch, maybe it got removed at some point. 

> if a training only accesses a small fraction of the data overall, that's fine too.
> 
> if it's random or not is another story, wouldn't matter here either way

for sure, that would be for a next iteration.

Reply:
ok as you prefer then, could remove it everywhere, or keep the gpt exception (maybe add another comment there that epochs are not used.) is epochs actually in the task configs, or do we have only rounds and steps per round there?

Reply:
added [an issue](https://github.com/epfml/disco/issues/689) as to not loss the discussion.

> is epochs actually in the task configs, or do we have only rounds and steps per round there?

in `Task`, we have `epochs` (total number of run over the whole dataset) & `roundDuration` (number of epochs before performing a round). steps are internal to gpt-tfjs (but should be flatten to reflect the `roundDuration` IMO).

> ok as you prefer then, could remove it everywhere, or keep the gpt exception (maybe add another comment there that epochs are not used.)

I was thinking of doing it after the v3 release but if you think it needs to be done before, I can get right to it.